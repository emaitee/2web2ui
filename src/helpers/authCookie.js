import cookie from 'js-cookie';
import config from 'src/config';

const { name, options } = config.authentication.cookie;

function save(data) {
  cookie.set(name, data, options);
}

function get() {
  return cookie.getJSON(name);
}

function remove() {
  return cookie.remove(name, options);
}

export default { save, get, remove };
