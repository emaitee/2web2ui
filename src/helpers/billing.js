import _ from 'lodash';

export function formatDataForCors(values) {
  const {
    firstName,
    lastName,
    cardType,
    cardNumber,
    expirationMonth,
    expirationYear,
    cardHolderName,
    securityCode,
    email,
    address1,
    address2,
    city,
    state,
    country,
    zipCode,
    selectedPlan
  } = values;

  // For CORS Endpoint + sift
  const corsData = {
    email,
    cardholder_name: cardHolderName,
    address1,
    address2,
    city,
    state,
    country,
    zip_code: zipCode,
    bin: cardNumber.slice(0, 6),
    last_four: cardNumber.slice(-4),
    plan_id: selectedPlan.billingId
  };

  // For Zuora
  const billingData = {
    billingId: selectedPlan.billingId,
    billToContact: {
      firstName,
      lastName,
      workEmail: email,
      country,
      state
    },
    creditCard: {
      cardType,
      cardNumber,
      expirationMonth,
      expirationYear,
      securityCode,
      cardHolderInfo: {
        cardHolderName: cardHolderName,
        addressLine1: address1,
        addressLine2: address2,
        city,
        zipCode
      }
    }
  };

  return { corsData, billingData };
}

export function formatCreateData({
  accountNumber,
  crmId,
  name,
  contractEffectiveDate,
  discountId = false,
  billingId,
  creditCard,
  billToContact
}) {
  const formatted = {
    accountNumber,
    autoPay: true,
    crmId,
    currency: 'USD',
    invoiceCollect: true,
    name,
    subscription: {
      contractEffectiveDate,
      subscribeToRatePlans: [{ productRatePlanId: billingId }],
      termType: 'EVERGREEN'
    },
    creditCard,
    billToContact
  };

  if (discountId) {
    formatted.subscription.subscribeToRatePlans.push({ productRatePlanId: discountId });
  }

  return formatted;
}

// Formats countries before storing in state
export function formatCountries(countries) {
  const ordered = _.flatten([
    _.remove(countries, { code: 'US' }),
    _.remove(countries, { code: 'GB' }),
    _.remove(countries, { code: 'CA' }),
    countries
  ]);

  return ordered.map((country) => formatForSelect(country));
}

function formatForSelect({ code, name, states }) {
  if (states) {
    return { value: code, label: name, states: states.map((state) => formatForSelect(state)) };
  }
  return { value: code, label: name };
}

export function getZipLabel(country) {
  if (country === 'US') {
    return 'Zip Code';
  }

  if (country === 'CA') {
    return 'Postal Code';
  }

  return 'Zip/Postal Code';
}
