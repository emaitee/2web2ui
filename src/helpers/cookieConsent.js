import * as jsCookie from 'js-cookie';
import config from 'src/config';

const { name, ageDays, options } = config.cookieConsent.cookie;

const isCookieSet = () => jsCookie.get(name) !== undefined;

const setCookie = () => jsCookie.set(name, '', { ...options, expires: ageDays });

export default { isCookieSet, setCookie };
