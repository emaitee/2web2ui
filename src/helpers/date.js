import moment from 'moment';

const relativeDateOptions = [
  { value: 'hour', label: 'Last Hour' },
  { value: 'day', label: 'Last 24 Hours' },
  { value: '7days', label: 'Last 7 Days' },
  { value: '30days', label: 'Last 30 Days' },
  { value: '90days', label: 'Last 90 Days' },
  { value: 'custom', label: 'Custom' }
];

export {
  getEndOfDay,
  getStartOfDay,
  getRelativeDates,
  relativeDateOptions
};

function getEndOfDay(date) {
  const end = new Date(date);
  end.setHours(23);
  end.setMinutes(59);
  end.setSeconds(59);
  end.setMilliseconds(0);

  return end;
}

function getStartOfDay(date) {
  const start = new Date(date);
  start.setHours(0);
  start.setMinutes(0);
  start.setSeconds(0);
  start.setMilliseconds(0);

  return start;
}

function getRelativeDates(range) {
  const now = moment.utc();
  const to = now.toDate();

  switch (range) {
    case 'hour':
      return { to, from: moment(to).subtract(1, 'hour').toDate() };

    case 'day':
      return { to, from: moment(to).subtract(1, 'day').toDate() };

    case '7days':
      return { to, from: moment(to).subtract(7, 'day').toDate() };

    case '30days':
      return { to, from: moment(to).subtract(30, 'day').toDate() };

    case '90days':
      return { to, from: moment(to).subtract(90, 'day').toDate() };

    default:
      return {};
  }
}
