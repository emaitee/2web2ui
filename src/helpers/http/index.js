export * from './sparkpostApi';
export { default as sparkpostLogin } from './sparkpostLogin';
export { default as sparkpostRefresh } from './sparkpostRefresh';
