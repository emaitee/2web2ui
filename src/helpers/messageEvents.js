/* eslint-disable */
import qs from 'query-string';
import _ from 'lodash';

/**
 * Reshapes message event documentation for tooltips
 */
export function formatDocumentation(data) {
  const events = {};

  _.each(data, (event) => {
    const { type, ...rest } = event;
    events[event.type.sampleValue] = _.mapValues(rest, ({ description }) => description);
  });

  return events;
}

export function parseSearch(search) {
  const { from, to, range, ...rest } = qs.parse(search);
  let dateOptions = {};

  if (from) {
    dateOptions.from = new Date(from);
  }

  if (to) {
    dateOptions.to = new Date(to);
  }

  if (range) {
    dateOptions.relativeRange = range;
  }

  const options = _.mapValues(rest, (filter, key) => {
    return typeof filter === 'string' ? [filter] : filter
  });

  return { dateOptions, ...options };
}
