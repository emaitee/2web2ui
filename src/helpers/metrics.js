import moment from 'moment';
import _ from 'lodash';
import { list as METRICS_LIST } from 'src/config/metrics';
import config from 'src/config';

const { metricsPrecisionMap: precisionMap, apiDateFormat, chartColors } = config;
const indexedPrecisions = _.keyBy(precisionMap, 'value');
const FILTER_KEY_MAP = {
  'Domain': 'domains',
  'Campaign': 'campaigns',
  'Template': 'templates',
  'Sending IP': 'sending_ips',
  'IP Pool': 'ip_pools',
  'Subaccount': 'subaccounts',
  'Sending/Bounce Domain': 'sending_domains'
};
const DELIMITERS = ',;:+~`!@#$%^*()-={}[]"\'<>?./|\\'.split('');

export function getQueryFromOptions({ from, to, metrics, activeList = []}) {
  from = moment(from).utc();
  to = moment(to).utc();

  const apiMetricsKeys = getKeysFromMetrics(metrics);
  const delimiter = getDelimiter(activeList);
  const filters = getFilterSets(activeList, delimiter);

  return {
    metrics: apiMetricsKeys.join(delimiter),
    precision: getPrecision(from, to),
    from: from.format(apiDateFormat),
    to: to.format(apiDateFormat),
    delimiter,
    ...filters
  };
}

export function pushToKey(obj, key, value) {
  const updated = { [key]: [], ...obj };
  updated[key].push(value);
  return updated;
}

export function getFilterSets(filters = [], delimiter) {
  const hash = filters.reduce((result, { type, value }) => pushToKey(result, FILTER_KEY_MAP[type], value), {});
  return _.mapValues(hash, (v) => v.join(delimiter));
}

/**
 * Get an array of unique chars based on all filters and
 * find all delimiters that don't appear in that list, then
 * return the first one to use as the delimiter
 *
 * @param {array} filters
 */
export function getDelimiter(filters = []) {
  const uniques = _.uniq(filters.map((f) => f.value).join(''));
  return _.difference(DELIMITERS, uniques).shift();
}

/**
 * Calculates the difference between 2 moment dates
 * and returns the closest precision value
 *
 */
export function getPrecision(from, to = moment()) {
  const diff = to.diff(from, 'minutes');
  return precisionMap.find(({ time }) => diff <= time).value;
}

export function getPrecisionType(precision) {
  return (indexedPrecisions[precision].time <= (60 * 24 * 2)) ? 'hours' : 'days';
}

export function getMetricsFromKeys(keys = []) {
  return keys.map((metric, i) => {
    const found = METRICS_LIST.find((M) => M.key === metric);
    return { ...found, name: found.key, stroke: chartColors[i] };
  });
}

export function getKeysFromMetrics(metrics = []) {
  const flattened = _.flatMap(metrics, ({ key, computeKeys }) => computeKeys ? computeKeys : key);
  return _.uniq(flattened);
}

export function computeKeysForItem(metrics = []) {
  return (item) => metrics.reduce((acc, metric) => {
    if (metric.compute) {
      acc[metric.key] = metric.compute(acc, metric.computeKeys) || 0;
    }
    return acc;
  }, item);
}

/**
 * Transforms API result into chart-ready data in 2 steps:
 * 1. compute any necessary computed metrics
 * 2. arrange metrics into groups sorted by unit/measure
 *
 * @param {Array} metrics - list of currently selected metrics objects from config
 * @param {Array} data - results from the metrics API
 */
export function transformData(data = [], metrics = []) {
  return data.map(computeKeysForItem(metrics));
}
