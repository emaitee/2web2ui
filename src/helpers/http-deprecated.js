import config from '../config';

import { sparkpost as sparkpostRequest } from 'src/helpers/axiosInstances';

const { authentication } = config;

// TODO handle timeout error better

function useRefreshToken(refreshToken) {
  return sparkpostRequest({
    method: 'POST',
    url: '/authenticate',
    data: `grant_type=refresh_token&refresh_token=${refreshToken}`,
    headers: authentication.headers
  });
}

function sparkpostLogin(username, password, rememberMe) {
  username = encodeURIComponent(username);
  password = encodeURIComponent(password);
  const data = `grant_type=password&username=${username}&password=${password}&rememberMe=${rememberMe}`;

  return sparkpostRequest({
    method: 'POST',
    url: '/authenticate',
    data,
    headers: authentication.headers
  });
}

export { sparkpostRequest, sparkpostLogin, useRefreshToken };
