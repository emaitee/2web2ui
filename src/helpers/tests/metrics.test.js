import * as metricsHelpers from '../metrics';
import moment from 'moment';
import * as dateHelpers from 'src/helpers/date';

describe('metrics helpers', () => {
  it('should build return correct options from updates', () => {
    const actual = metricsHelpers.getQueryFromOptions({
      from: '2017-12-18T00:00',
      to: '2017-12-18T11:00',
      metrics: [
        { key: 'count_bounce' },
        { key: 'foo_bar', computeKeys: 'test + last' }
      ],
      activeList: [
        { value: 'gmail.com', type: 'Recipient Domain' },
        { value: 'foobar', type: 'Subaccount', id: 100 }
      ]
    });

    expect(actual).toMatchSnapshot();
  });

  it('should pushToKey', () => {
    const actual = metricsHelpers.pushToKey({ foo: 'bar', baz: 'bot' }, 'test', 'value');
    expect(actual).toMatchSnapshot();
  });

  it('should set appropriate delimiter', () => {
    const actual = metricsHelpers.getDelimiter([
      { value: 'test,me' },
      { value: 'foo;bar' },
      { value: 'test:me' }
    ]);

    expect(actual).toEqual('+');
  });

  it('should getPrecision from dates', () => {
    const from = moment('2016-12-18T00:00').utc();
    const to = moment('2016-12-18T00:30').utc();

    expect(metricsHelpers.getPrecision(from, to)).toEqual('1min');

    to.add(1, 'hours');
    expect(metricsHelpers.getPrecision(from, to)).toEqual('5min');

    to.add(2, 'hours');
    expect(metricsHelpers.getPrecision(from, to)).toEqual('15min');

    to.add(5, 'hours');
    expect(metricsHelpers.getPrecision(from, to)).toEqual('hour');

    to.add(6, 'days');
    expect(metricsHelpers.getPrecision(from, to)).toEqual('12hr');

    to.add(25, 'days');
    expect(metricsHelpers.getPrecision(from, to)).toEqual('day');

    to.add(30, 'days');
    expect(metricsHelpers.getPrecision(from, to)).toEqual('week');

    to.add(200, 'days');
    expect(metricsHelpers.getPrecision(from, to)).toEqual('month');

  });

  it('should return hours as precision type', () => {
    expect(metricsHelpers.getPrecisionType('1min')).toEqual('hours');
  });

  it('should return days as precision type', () => {
    expect(metricsHelpers.getPrecisionType('12hr')).toEqual('days');
  });

  it('should get metrics from keys', () => {
    const actual = metricsHelpers.getMetricsFromKeys(['count_delayed', 'count_injected']);
    expect(actual).toMatchSnapshot();
  });

  it('should compute keys during transform', () => {
    const data = [
      { count_rendered: 100, count_unique_confirmed_opened_approx: 1, count_targeted: 100 },
      { count_rendered: 101 }
    ];

    const metrics = [
      { key: 'count_rendered' },
      { key: 'open_rate_approx', computeKeys: ['count_unique_confirmed_opened_approx', 'count_targeted'], type: 'percentage', compute: (item, keys) => ((item[keys[0]] / item[keys[1]]) * 100) }
    ];

    const actual = metricsHelpers.transformData(data, metrics);
    expect(actual).toMatchSnapshot();
  });

  it('should build common metric options', () => {
    dateHelpers.getRelativeDates = jest.fn(() => ({ from: 'foo', to: 'bar' }));

    const actual = metricsHelpers.buildCommonOptions({ domain: 'gmail' }, { campaign: 'test', relativeRange: '7days' });

    expect(actual).toMatchSnapshot();

  });

});
