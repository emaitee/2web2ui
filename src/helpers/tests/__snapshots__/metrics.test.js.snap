// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metrics helpers should build common metric options 1`] = `
Object {
  "campaign": "test",
  "domain": "gmail",
  "from": "foo",
  "relativeRange": "7days",
  "to": "bar",
}
`;

exports[`metrics helpers should build return correct options from updates 1`] = `
Object {
  "delimiter": ",",
  "domains": "gmail.com",
  "from": "2017-12-18T05:00",
  "metrics": "count_bounce,test + last",
  "precision": "hour",
  "subaccounts": "100",
  "to": "2017-12-18T16:00",
}
`;

exports[`metrics helpers should compute keys during transform 1`] = `
Array [
  Object {
    "count_rendered": 100,
    "count_targeted": 100,
    "count_unique_confirmed_opened_approx": 1,
    "open_rate_approx": 1,
  },
  Object {
    "count_rendered": 101,
    "open_rate_approx": 0,
  },
]
`;

exports[`metrics helpers should get metrics from keys 1`] = `
Array [
  Object {
    "description": "Total number of delays due to any temporary failure.",
    "inSummary": true,
    "key": "count_delayed",
    "label": "Delayed",
    "name": "count_delayed",
    "stroke": "#04AEF9",
    "type": "total",
    "unit": "number",
  },
  Object {
    "description": "Messages either injected to or received by SparkPost.",
    "inSummary": true,
    "key": "count_injected",
    "label": "Injected",
    "name": "count_injected",
    "stroke": "#fa6423",
    "type": "total",
    "unit": "number",
  },
]
`;

exports[`metrics helpers should pushToKey 1`] = `
Object {
  "baz": "bot",
  "foo": "bar",
  "test": Array [
    "value",
  ],
}
`;
